pennylane/devices/qubit/preprocess.py

```diff
@@ -77,6 +78,76 @@ def _operator_decomposition_gen(
 #######################
 
 
+def validate_and_expand_adjoint(
+    circuit: qml.tape.QuantumScript,
+) -> qml.tape.QuantumScript:  # pylint: disable=protected-access
+    """Function for validating that the operations and observables present in the input circuit
+    are valid for adjoint differentiation.
+
+    Args:
+        circuit(.QuantumTape): the tape to validate
+
+    Returns:
+        .QuantumTape: The expanded tape, such that it is supported by adjoint differentiation
+
+    Raises:
+        QuantumFunctionError: if the circuit is invalid for adjoint differentiation
+    """
+    # Check validity of measurements
+    measurements = []
+    for m in circuit.measurements:
+        if not isinstance(m, ExpectationMP):
+            raise qml.QuantumFunctionError(
+                "Adjoint differentiation method does not support"
+                f" measurement {m.__class__.__name__}"
+            )
+
+        if m.obs.name in {"Hamiltonian"}:
+            raise qml.QuantumFunctionError(
+                f"Adjoint differentiation method does not support observable {m.obs.name}."
+            )
+
+        if not hasattr(m.obs, "base_name"):
+            m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        measurements.append(m)
+
+    expanded_ops = []
+    for op in circuit._ops:
+        if op.num_params > 1:
+            if not isinstance(op, qml.Rot):
```

Comment:
I know this is just a copy-paste from `_qubit_device`, but why don't we instead check that it decomposes into ops with 0-1 params instead? Do we actually want special support for Rot only?

Reply:
Or `op.has_generator`.

Reply:
Because `Rot` is the only multi-param operation where the parameters of the decomposed ops are the same as the parameters of the operation itself. Example:

```
>>> params = (0.1, 0.2, 0.3)
>>> qml.CRot(*params, wires=[0, 1]).decomposition()
[RZ(-0.09999999999999999, wires=[1]), CNOT(wires=[0, 1]), RZ(-0.2, wires=[1]), RY(-0.1, wires=[1]), CNOT(wires=[0, 1]), RY(0.1, wires=[1]), RZ(0.3, wires=[1])]
```

Reply:
This code will be tracked by the machine-learning interface, so it can include classical transformations. We basically want to be able to remove [this patch](https://github.com/PennyLaneAI/pennylane/blob/2efbb3b07d3665a2a208bed777c9120940a47bdd/pennylane/interfaces/execution.py#L63) when using the new device.